File: learning_tf/src/turtle_tf_listener.cpp
#include <ros/ros.h>
#include <tf/transform_listener.h>
#include <geometry_msgs/Twist.h>
#include <turtlesim/Spawn.h>

int main(int argc, char ** argv) {
    ros::init(argc, argv, "ycao_tf_listener");

    ros::NodeHandle nh;

    ros::ServiceClient spawnClient = nh.serviceClient<turtlesim::Spawn>("/spawn");

    ros::service::waitForService("/spawn");
    
    turtlesim::Spawn::Request req;
    turtlesim::Spawn::Response resp;
    bool ok = spawnClient.call(req, resp);
    if (ok) {
        ROS_INFO("spawned a turtle named %s", resp.name.c_str());
    } else {
        ROS_ERROR("Failed to spawn");
    }

    ros::Publisher cmd_pub = nh.advertise<geometry_msgs::Twist>("/turtle2/cmd_vel", 10);

    tf::TransformListener listener;

    ros::Rate loop_rate(10);
    while (nh.ok()) {
        tf::StampedTransform tf_data;
        try {
			listener.waitForTransform("/turtle2", "/turtle1", ros::Time(0), ros::Duration(3.0));
			listener.lookupTransform("/turtle2", "/turtle1", ros::Time(0), tf_data);            
        } catch (tf::TransformException &ex) {
            ROS_ERROR("%s", ex.what());
            ros::Duration(1.0).sleep();
            continue;
        }

        geometry_msgs::Twist vel;
        vel.linear.x = 0.5 * sqrt(pow(tf_data.getOrigin().x(), 2) + pow(tf_data.getOrigin().y(), 2));
        vel.angular.z = 4.0 * atan2(tf_data.getOrigin().y(), tf_data.getOrigin().x());

        // 下面的改动，可以让turtle1控制turtle2的转向，即turtle2跟随turtle1转，但turtle2追不上turtle1
        // double roll, pitch, yaw;
        // tf::Quaternion qua = tf_data.getRotation();
        // tf::Matrix3x3(qua).getRPY(roll, pitch, yaw);
        // vel.angular.z = 4.0 * yaw;

        cmd_pub.publish(vel);

        loop_rate.sleep();
        ros::spinOnce();
    }

    return 0;
}

File: learning_tf/scripts/turtle_tf_listener.py
#！/usr/bin/env python3

import roslib
import rospy
import math
import tf

from geometry_msgs.msg import Twist
from turtlesim.srv import Spawn

def main():
  rospy.init_node("turtle_tf_listener")
  
  rospy.wait_for_service("spawn")
  client = rospy.ServiceProxy("spawn", Spawn)
  
  try:
    req = SpawnRequest()
    resp = client(req)
    rospy.loginfo("spawned a turtle name %s" %resp.name)
  except rospy.ServiceException as e:
      rospy.loginfo("service call fail: %s" %e)

  cmd_pub= rospy.Publisher("/turtle2/cmd_vel", Twist, queue_size=10)
  
  tf_listener = tf.TransformListener()
  
  rate = rospy.Rate(10)
  while not rospy.is_shutdown():
    try:
      (translation, rotation) = tf_listener.lookupTransform("/turtle2", "turtle1", rospy.Time(0))
    except (tf.LookupException, tf.ConnectivityException, tf.ExtrapolationException):
      continue

    cmd_vel = Twist()
    cmd_vel.linear.x= 0.5 * math.sqrt(trans[0] ** 2 + trans[1] ** 2)
    cmd_vel.angular.z = 4 * math.atan2(translation[1], translation[0])

    # 下面的改动，可以让turtle1控制turtle2的转向，即turtle2跟随turtle1转，但turtle2追不上turtle1
    # quaternion = rotation
    # # euler = (roll, pitch, yaw)
    # euler = tf.transformations.euler_from_quaternion(quaternion)
    # cmd_vel.angular.z = 4 * euler[2]
    
    cmd_pub.publish(cmd_vel)
    
    rate.sleep()

if __name__ == "__main__":
  main()